org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/utils/CompletionItemDefaultsUtils.java

```diff
@@ -0,0 +1,115 @@
+/*******************************************************************************
+* Copyright (c) 2023 Red Hat Inc. and others.
+* All rights reserved. This program and the accompanying materials
+* which accompanies this distribution, and is available at
+* http://www.eclipse.org/legal/epl-v20.html
+*
+* SPDX-License-Identifier: EPL-2.0
+*
+* Contributors:
+*     Red Hat Inc. - initial API and implementation
+*******************************************************************************/
+package org.eclipse.lemminx.utils;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.eclipse.lemminx.services.CompletionResponse;
+import org.eclipse.lemminx.settings.SharedSettings;
+import org.eclipse.lsp4j.CompletionItem;
+import org.eclipse.lsp4j.CompletionItemDefaults;
+import org.eclipse.lsp4j.InsertTextFormat;
+import org.eclipse.lsp4j.Range;
+import org.eclipse.lsp4j.jsonrpc.messages.Either;
+
+/**
+ * CompletionItemDefaultsUtils for implementing itemDefaults in completion list
+ */
+public class CompletionItemDefaultsUtils {
+
+	private static final String ITEM_DEFAULTS_EDIT_RANGE = "editRange";
+
+	private static final String ITEM_DEFAULTS_INSERT_TEXT_FORMAT = "insertTextFormat";
+
+	public static void process(CompletionResponse completionResponse, SharedSettings sharedSettings) {
+		CompletionItemDefaults itemDefaults = new CompletionItemDefaults();
+		List<CompletionItem> completionList = completionResponse.getItems();
+		if (sharedSettings.getCompletionSettings().isCompletionListItemDefaultsSupport(ITEM_DEFAULTS_EDIT_RANGE)) {
+			Range editRange = findMostCommonEditRange(completionList);
+			itemDefaults.setEditRange(Either.forLeft(editRange));
+			for (CompletionItem item : completionList) {
+				if (item.getTextEdit().isLeft() && item.getTextEdit().getLeft().getRange().equals(editRange)) {
+					item.setTextEditText(item.getTextEdit().getLeft().getNewText());
+					item.setTextEdit(null);
+				}
+			}
+			completionResponse.setItemDefaults(itemDefaults);
+		}
+		if (sharedSettings.getCompletionSettings()
+				.isCompletionListItemDefaultsSupport(ITEM_DEFAULTS_INSERT_TEXT_FORMAT)) {
+			itemDefaults.setInsertTextFormat(findMostCommonInsertTextFormat(completionResponse.getItems()));
+			for (CompletionItem item : completionList) {
+				if (item.getInsertTextFormat() == itemDefaults.getInsertTextFormat()) {
+					item.setInsertTextFormat(null);
+				}
+			}
+			completionResponse.setItemDefaults(itemDefaults);
+		}
+	}
+
+	/**
+	 * Returns the most common editRange in the completion list.
+	 *
+	 * @param completionList the completion response
+	 * @return the most common editRange in the completion list
+	 */
+	private static Range findMostCommonEditRange(List<CompletionItem> completionList) {
+		Map<Range, Integer> countMapEditRange = new HashMap<>();
```

Comment:
Instead of having a Map with Map<Range, Integer>, I wonder ifit should be beter to have a `Map<Range, List<CompletionItem>>`, it will avoid to loop of for `(CompletionItem item : completionList) {`  to update setInsertTextFormat to null.

Reply:
I'm not sure if I understand, how would it be able to count the number of occurrences of the range?

Reply:
perhaps I'm wrong, but my first impression that you count the number of item which have the same range. In order to store the count, my suggestion was to store a list of item which have the same range. The count number is the size list, no?

Reply:
Ohhh I see, will do that, thanks!